presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java

```diff
@@ -249,6 +254,43 @@ public static Domain getDomain(Type type, long rowCount, Statistics<?> statistic
             return createDomain(type, hasNullValue, parquetTimestampStatistics);
         }
 
+        if (type instanceof TimestampType && statistics instanceof LongStatistics) {
+            TimestampType timestampType = (TimestampType) type;
+            LongStatistics longStatistics = (LongStatistics) statistics;
+            if (longStatistics.genericGetMin() > longStatistics.genericGetMax()) {
+                failWithCorruptionException(failOnCorruptedParquetStatistics, column, id, longStatistics);
+                return Domain.create(ValueSet.all(type), hasNullValue);
+            }
+
+            long min = longStatistics.getMin();
+            long max = longStatistics.getMax();
+
+            TimeUnit timestampPrecision;
+            int prestoTimestampPrecision = timestampType.getPrecision();
+            if (prestoTimestampPrecision <= 3) {
+                timestampPrecision = MILLISECONDS;
+            }
+            else if (prestoTimestampPrecision <= 6) {
+                timestampPrecision = MICROSECONDS;
+            }
+            else {
+                throw new PrestoException(NOT_SUPPORTED, "Parquet does not support timestamp types with precision: " + statistics.type().getOriginalType());
+            }
+
+            ParquetIntegerStatistics parquetIntegerStatistics;
+            switch (statistics.type().getOriginalType()) {
+                case TIMESTAMP_MILLIS:
+                    parquetIntegerStatistics = new ParquetIntegerStatistics(timestampPrecision.convert(min, MILLISECONDS), timestampPrecision.convert(max, MILLISECONDS));
+                    break;
+                case TIMESTAMP_MICROS:
+                    parquetIntegerStatistics = new ParquetIntegerStatistics(timestampPrecision.convert(min, MICROSECONDS), timestampPrecision.convert(max, MICROSECONDS));
+                    break;
+                default:
+                    throw new PrestoException(NOT_SUPPORTED, "Found unsupported Parquet timestamp type: " + statistics.type().getOriginalType());
+            }
+            return createDomain(type, hasNullValue, parquetIntegerStatistics);
```

Comment:
Maybe something like:

```java
            TimeUnit prestoTimestampRepresentationUnit = prestoTimestampRepresentationUnit((TimestampType) type);
            TimeUnit parquetTimestampRepresentationUnit = parquetTimestampRepresentationUnit(statistics.type().getOriginalType());
            ParquetTimestampStatistics parquetTimestampStatistics = new ParquetTimestampStatistics(
                    prestoTimestampRepresentationUnit.convert(longStatistics.getMin(), parquetTimestampRepresentationUnit),
                    prestoTimestampRepresentationUnit.convert(longStatistics.getMax(), parquetTimestampRepresentationUnit));
            return createDomain(type, hasNullValue, parquetTimestampStatistics);
```

with

```java
private static TimeUnit prestoTimestampRepresentationUnit(TimestampType timestampType)
{
    if (timestampType.getPrecision() <= 3) {
        return MILLISECONDS;
    }
    if (timestampType.getPrecision() <= 6) {
        return MICROSECONDS;
    }
    throw new IllegalArgumentException("Unsupported type: " + timestampType);
}

private static TimeUnit parquetTimestampRepresentationUnit(OriginalType originalType)
{
    switch (originalType) {
        case TIMESTAMP_MILLIS:
            return MILLISECONDS;
        case TIMESTAMP_MICROS:
            return MICROSECONDS;
    }
    throw new IllegalArgumentException("Unsupported original type: " + originalType);
}
```

Reply:
Much better, thanks. I just put the type conversion methods into `ParquetTimestampUtils`